File: mobile/lib/features/attendance/domain/attendance_record.dart
class AttendanceRecord {
  const AttendanceRecord({
    required this.id,
    required this.tenantId,
    required this.userId,
    required this.date,
    required this.status,
    this.clockInAt,
    this.clockOutAt,
    this.clockInLatitude,
    this.clockInLongitude,
    this.clockOutLatitude,
    this.clockOutLongitude,
    this.clockInDeviceId,
    this.clockOutDeviceId,
    this.clockOutOutsideRadius = false,
    this.workDurationMinutes,
  });

  final String id;
  final String tenantId;
  final String userId;
  final String date;
  final String status;
  final DateTime? clockInAt;
  final DateTime? clockOutAt;
  final double? clockInLatitude;
  final double? clockInLongitude;
  final double? clockOutLatitude;
  final double? clockOutLongitude;
  final String? clockInDeviceId;
  final String? clockOutDeviceId;
  final bool clockOutOutsideRadius;
  final int? workDurationMinutes;

  bool get hasClockedIn => clockInAt != null;
  bool get hasClockedOut => clockOutAt != null;

  factory AttendanceRecord.fromJson(Map<String, dynamic> json) {
    return AttendanceRecord(
      id: json['id']?.toString() ?? '',
      tenantId: json['tenantId']?.toString() ?? '',
      userId: json['userId']?.toString() ?? '',
      date: json['date']?.toString() ?? '',
      status: json['status']?.toString() ?? 'UNKNOWN',
      clockInAt: _dateTime(json['clockInAt']),
      clockOutAt: _dateTime(json['clockOutAt']),
      clockInLatitude: _double(json['clockInLatitude']),
      clockInLongitude: _double(json['clockInLongitude']),
      clockOutLatitude: _double(json['clockOutLatitude']),
      clockOutLongitude: _double(json['clockOutLongitude']),
      clockInDeviceId: json['clockInDeviceId']?.toString(),
      clockOutDeviceId: json['clockOutDeviceId']?.toString(),
      clockOutOutsideRadius: json['clockOutOutsideRadius'] == true,
      workDurationMinutes: _int(json['workDurationMinutes']),
    );
  }

  static DateTime? _dateTime(Object? value) {
    if (value == null) return null;
    return DateTime.tryParse(value.toString());
  }

  static double? _double(Object? value) {
    if (value is num) return value.toDouble();
    return double.tryParse(value?.toString() ?? '');
  }

  static int? _int(Object? value) {
    if (value is num) return value.toInt();
    return int.tryParse(value?.toString() ?? '');
  }
}

File: backend/src/main/kotlin/com/hadivo/attendance/modules/attendance/AttendanceRecord.kt
package com.hadivo.attendance.modules.attendance

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.PreUpdate
import jakarta.persistence.Table
import java.time.Instant
import java.time.LocalDate
import java.time.LocalTime
import java.util.UUID

@Entity
@Table(name = "attendance_records")
class AttendanceRecord(
    @Column(name = "tenant_id", nullable = false)
    var tenantId: UUID,

    @Column(name = "user_id", nullable = false)
    var userId: UUID,

    @Column(nullable = false)
    var date: LocalDate,

    @Column(name = "clock_in_at")
    var clockInAt: Instant? = null,

    @Column(name = "clock_out_at")
    var clockOutAt: Instant? = null,

    @Column(name = "clock_in_location_id")
    var clockInLocationId: UUID? = null,

    @Column(name = "clock_out_location_id")
    var clockOutLocationId: UUID? = null,

    @Column(name = "clock_in_latitude")
    var clockInLatitude: Double? = null,

    @Column(name = "clock_in_longitude")
    var clockInLongitude: Double? = null,

    @Column(name = "clock_out_latitude")
    var clockOutLatitude: Double? = null,

    @Column(name = "clock_out_longitude")
    var clockOutLongitude: Double? = null,

    @Column(name = "clock_in_device_id")
    var clockInDeviceId: String? = null,

    @Column(name = "clock_out_device_id")
    var clockOutDeviceId: String? = null,

    @Column(name = "clock_out_outside_radius", nullable = false)
    var clockOutOutsideRadius: Boolean = false,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var status: AttendanceStatus,

    @Column(name = "work_duration_minutes")
    var workDurationMinutes: Int? = null,

    @Column(name = "shift_template_id")
    var shiftTemplateId: UUID? = null,

    @Column(name = "shift_name", length = 120)
    var shiftName: String? = null,

    @Column(name = "scheduled_start_time")
    var scheduledStartTime: LocalTime? = null,

    @Column(name = "scheduled_end_time")
    var scheduledEndTime: LocalTime? = null,

    @Column(name = "late_threshold_minutes")
    var lateThresholdMinutes: Int? = null,
) {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    var id: UUID? = null

    @Column(name = "created_at", nullable = false, updatable = false)
    var createdAt: Instant = Instant.now()

    @Column(name = "updated_at", nullable = false)
    var updatedAt: Instant = Instant.now()

    @PreUpdate
    fun touchUpdatedAt() {
        updatedAt = Instant.now()
    }
}
